sql/transform/expr.go

```diff
@@ -12,60 +12,68 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-package expression
+package transform
 
 import (
 	"errors"
 
 	"github.com/dolthub/go-mysql-server/sql"
 )
 
-// TransformExprWithNodeFunc is a function that given an expression and the node that contains it, will return that
-// expression as is or transformed along with an error, if any.
-type TransformExprWithNodeFunc func(sql.Node, sql.Expression) (sql.Expression, error)
-
-// TransformUp applies a transformation function to the given expression from the
-// bottom up.
-func TransformUp(e sql.Expression, f sql.TransformExprFunc) (sql.Expression, error) {
+// Expr applies a transformation function to the given expression tree from the bottom up.
+func Expr(e sql.Expression, f ExprFunc) (sql.Expression, TreeIdentity, error) {
 	children := e.Children()
 	if len(children) == 0 {
-		e, err := e.WithChildren()
+		return f(e)
+	}
+
+	var (
+		newChildren []sql.Expression
+		c           sql.Expression
+		sameC       = SameTree
+		err         error
+	)
+
+	for i := 0; i < len(children); i++ {
+		c = children[i]
+		c, sameC, err = Expr(c, f)
 		if err != nil {
-			return nil, err
+			return nil, SameTree, err
+		}
+		if !sameC {
+			if newChildren == nil {
+				newChildren = make([]sql.Expression, len(children))
+				copy(newChildren, children)
+			}
+			newChildren[i] = c
 		}
-		return f(e)
 	}
 
-	newChildren := make([]sql.Expression, len(children))
-	for i, c := range children {
-		c, err := TransformUp(c, f)
+	if len(newChildren) > 0 {
+		sameC = NewTree
```

Comment:
This is actually a semantically different var than the one inside the loop (which is logically scoped to a single interation)